File: Linked_List/reverse.cpp
#include<bits/stdc++.h>
using namespace std;

// Linked List Node 
struct Node {
	int data;
	struct Node* next;
	Node(int data) {
		this->data = data;
		next = NULL;
	}
};

struct LinkedList {
	Node* head;
	LinkedList() { head = NULL; }

	void print() {
		struct Node* temp = head;
		while (temp != NULL) {
			cout << temp->data << " ";
			temp = temp->next;
		}
	}

	void push(int data) {
		Node* temp = new Node(data);
		temp->next = head;
		head = temp;
	}

	void reverse(){
		Node* curr = head;
		Node* prev = NULL;
		Node* temp = NULL;

		while(curr != NULL){
			temp = curr->next;
			curr->next = prev;
			prev = curr;
			curr = temp;
		}
		head = prev;
	}
};


int main() {
	LinkedList LL;
	LL.push(1);
	LL.push(2);
	LL.push(3);
	LL.push(4);
	LL.push(5);

	LL.print();
	cout << "\n";
	LL.reverse();
	LL.print();
}

File: Linked_List/reverse.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
class LinkedList:
    def __init__(self):
        self.head = None
        
        
    def push(self, data):
        new_node = Node(data)
        new_node.next = self.head 
        self.head = new_node
    
    def printList(self):
        temp = self.head
        while(temp):
            print(temp.data, end = " ")
            temp = temp.next

    def reverse(self):
        prev = None
        curr = self.head 
        while (curr is not None):
            temp = curr.next 
            curr.next =prev 
            prev = curr 
            curr = temp 
        self.head = prev 
            
LL = LinkedList()
LL.push(1)
LL.push(2)
LL.push(3)
LL.push(4)
LL.printList()
print("\n")
LL.reverse()
LL.printList()
